jina/types/arrays/document.py

```diff
@@ -397,18 +406,28 @@ def overriden_key(proto):
             # sure (quite fragile) the `sort` will work seamlessly, it tries to apply `key` to the first element and
             # see if it works. If it works it can sort with `proto` interface, otherwise use `Document` interface one.
             # (Very often the 2 interfaces are both the same and valid, so proto will have less overhead
-            overriden = False
+            _key = key
             try:
                 key(self._pb_body[0])
             except:
-                overriden = True
+                _key = overriden_key
 
-            if not overriden:
-                self._pb_body.sort(key=key, *args, **kwargs)
-            else:
-                self._pb_body.sort(key=overriden_key, *args, **kwargs)
+        if top_k is None or top_k >= len(self._pb_body):
+            self._pb_body.sort(key=_key, reverse=reverse)
         else:
-            self._pb_body.sort(*args, **kwargs)
+            if _key is not None:
+                self._pb_body = [(_key(element), element) for element in self._pb_body]
```

Comment:
I would prefer not to change `self._pb_body`, I see that u do too to avoid having the `if` else at the other level. But all the magic done later seems to be more confusing. I would separate quite clearly what happens to the actual `_pb_body` and what are the values put in the heap.

Reply:
Also, if some exception happens after here, the DocumentArray will stay in a very invalid state